Source/Scene/PointCloud.js

```diff
@@ -0,0 +1,1315 @@
+        }
+
+        var usesColorSemantic = (styleableProperties.indexOf('COLOR') >= 0);
+        var usesNormalSemantic = (styleableProperties.indexOf('NORMAL') >= 0);
+
+        // Split default properties from user properties
+        var userProperties = styleableProperties.filter(function(property) { return defaultProperties.indexOf(property) === -1; });
+
+        if (usesNormalSemantic && !hasNormals) {
+            throw new RuntimeError('Style references the NORMAL semantic but the point cloud does not have normals');
+        }
+
+        // Disable vertex attributes that aren't used in the style, enable attributes that are
+        var styleableShaderAttributes = pointCloud._styleableShaderAttributes;
+        for (name in styleableShaderAttributes) {
+            if (styleableShaderAttributes.hasOwnProperty(name)) {
+                attribute = styleableShaderAttributes[name];
+                var enabled = (userProperties.indexOf(name) >= 0);
+                var vertexAttribute = getVertexAttribute(vertexArray, attribute.location);
+                vertexAttribute.enabled = enabled;
+            }
+        }
+
+        var usesColors = hasColors && (!hasColorStyle || usesColorSemantic);
+        if (hasColors) {
+            // Disable the color vertex attribute if the color style does not reference the color semantic
+            var colorVertexAttribute = getVertexAttribute(vertexArray, colorLocation);
+            colorVertexAttribute.enabled = usesColors;
+        }
+
+        var attributeLocations = {
+            a_position : positionLocation
+        };
+        if (usesColors) {
+            attributeLocations.a_color = colorLocation;
+        }
+        if (hasNormals) {
+            attributeLocations.a_normal = normalLocation;
+        }
+        if (hasBatchIds) {
+            attributeLocations.a_batchId = batchIdLocation;
+        }
+
+        var attributeDeclarations = '';
+
+        var length = userProperties.length;
+        for (i = 0; i < length; ++i) {
+            name = userProperties[i];
+            attribute = styleableShaderAttributes[name];
+            if (!defined(attribute)) {
+                throw new RuntimeError('Style references a property "' + name + '" that does not exist or is not styleable.');
+            }
+
+            var componentCount = attribute.componentCount;
+            var attributeName = 'czm_tiles3d_style_' + name;
+            var attributeType;
+            if (componentCount === 1) {
+                attributeType = 'float';
+            } else {
+                attributeType = 'vec' + componentCount;
+            }
+
+            attributeDeclarations += 'attribute ' + attributeType + ' ' + attributeName + '; \n';
+            attributeLocations[attributeName] = attribute.location;
+        }
+
+        createUniformMap(pointCloud, frameState);
+
+        var vs = 'attribute vec3 a_position; \n' +
```

Comment:
For the current use case, how does the shader cache perform?  It doesn't thrash right - since a shader program isn't deleted right away?  Nonetheless it has to do a ton of string processing - could be a good roadmap or `PERFORMANCE_IDEA` to improve just like reusing a vertex buffer.

Reply:
Yeah, no shader cache thrashing but there is a bunch of string processing. This and other GPU resource optimizations are now in a `PERFORMANCE_IDEA`.